### PR TITLE
SONAR-10163 updating built-in Rule now returns HTTP 400

### PR DESCRIPTION
**Pull request by gjambet**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/rule/RuleUpdate.java [hunk 1]
```diff
@@ -224,7 +224,7 @@ private boolean isCustomRuleFieldsEmpty() {
 
   private void checkCustomRule() {
     if (useCase != CUSTOM_RULE) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/rule/RuleUpdate.java]: Please use Preconditions.checkArgument

File: server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java [hunk 2]
```diff
@@ -82,24 +86,22 @@
...
   private RuleIndexer ruleIndexer = new RuleIndexer(es.client(), db.getDbClient());
   private DbSession dbSession = db.getSession();
-  private TestDefaultOrganizationProvider defaultOrganizationProvider = TestDefaultOrganizationProvider.from(db);
 
   private RuleUpdater underTest = new RuleUpdater(db.getDbClient(), ruleIndexer, system2);
 
   @Test
   public void do_not_update_rule_with_removed_status() {
-    db.rules().insert(RuleTesting.newRule(RULE_KEY).setStatus(RuleStatus.REMOVED));
+    db.rules().insert(RuleTesting.newRule(RULE_KEY).setStatus(REMOVED));
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java]: Please use consumer, it's more easily readable : 
```
db.rules().insert(rule -> rule.setStatus(REMOVED)
```

Comment (julienlancelot) [reply server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java]: Moreover, no need to specify a key, using a random one should be enough

File: server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java [hunk 3]
```diff
@@ -512,12 +514,11 @@ public void fail_to_update_custom_rule_when_empty_name() {
...
       .setMarkdownDescription("New desc")
       .setOrganization(db.getDefaultOrganization());
-    try {
-      underTest.update(dbSession, update, db.getDefaultOrganization(), userSessionRule);
-      fail();
-    } catch (Exception e) {
-      assertThat(e).isInstanceOf(IllegalArgumentException.class).hasMessage("The name is missing");
-    }
+
+    expectedException.expect(IllegalArgumentException.class);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java]: It would have be easier to review if the refactoring to use an ExpectedException would have been done in another commit

File: server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java [hunk 4]
```diff
@@ -537,66 +538,49 @@ public void fail_to_update_custom_rule_when_empty_description() {
...
+    expectedException.expectMessage("The description is missing");
+
+    underTest.update(dbSession, update, db.getDefaultOrganization(), userSessionRule);
   }
 
   @Test
   public void fail_to_update_plugin_rule_if_name_is_set() {
     // Create rule rule
     RuleDefinitionDto ruleDto = RuleTesting.newRule(RuleKey.of("squid", "S01"));
     db.rules().insert(ruleDto);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java]: Please inline the 2 previous line

File: server/sonar-server/src/test/java/org/sonar/server/rule/ws/UpdateActionTest.java [hunk 5]
```diff
@@ -310,6 +310,27 @@ public void fail_to_update_custom_when_description_is_empty() {
       .execute();
   }
 
+  @Test
+  public void throw_IllegalArgumentException_if_trying_to_update_builtin_rule_description() throws Exception {
+    logInAsQProfileAdministrator();
+    RuleDefinitionDto templateRule = db.rules().insert(
+      r -> r.setRuleKey(RuleKey.of("java", "S001")),
+      r -> r.setIsTemplate(true),
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/rule/ws/UpdateActionTest.java]: A template rule is not necessarily a built-in rule.
Please simply the creation of the rule by only using `db.rules().insert()`, not need to override anything

File: server/sonar-server/src/main/java/org/sonar/server/rule/RuleUpdate.java (hunk 1 above)
Comment (gjambet) [reply server/sonar-server/src/main/java/org/sonar/server/rule/RuleUpdate.java]: done

File: server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java (hunk 4 above)
Comment (gjambet) [reply server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java]: done

File: server/sonar-server/src/test/java/org/sonar/server/rule/ws/UpdateActionTest.java (hunk 5 above)
Comment (gjambet) [reply server/sonar-server/src/test/java/org/sonar/server/rule/ws/UpdateActionTest.java]: done

File: server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java (hunk 3 above)
Comment (gjambet) [reply server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java]: agreed

File: server/sonar-server/src/main/java/org/sonar/server/rule/RuleUpdate.java [hunk 6]
```diff
@@ -19,6 +19,7 @@
  */
 package org.sonar.server.rule;
 
+import com.google.common.base.Preconditions;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/rule/RuleUpdate.java]: MINOR Remove this unused import 'com.google.common.base.Preconditions'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java (hunk 4 above)
Comment (julienlancelot) [reply server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java]: Done where ? I don't see it :)

File: server/sonar-server/src/test/java/org/sonar/server/rule/ws/UpdateActionTest.java [hunk 7]
```diff
@@ -310,6 +310,22 @@ public void fail_to_update_custom_when_description_is_empty() {
       .execute();
   }
 
+  @Test
+  public void throw_IllegalArgumentException_if_trying_to_update_builtin_rule_description() throws Exception {
+    logInAsQProfileAdministrator();
+    RuleDefinitionDto templateRule = db.rules().insert();
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/rule/ws/UpdateActionTest.java]: name of the variable is not "templateRule" but "builtInRule" or only "rule"

File: server/sonar-server/src/test/java/org/sonar/server/rule/ws/UpdateActionTest.java [hunk 8]
```diff
@@ -310,6 +310,22 @@ public void fail_to_update_custom_when_description_is_empty() {
...
+
+    expectedException.expect(IllegalArgumentException.class);
+    expectedException.expectMessage("Not a custom rule");
+
+    ws.newRequest().setMethod("POST")
+      .setParam("key", templateRule.getKey().toString())
+      .setParam("name", "My custom rule")
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/rule/ws/UpdateActionTest.java]: "My custom rule" is misleading as it suggest your updating a custom rule.
Please use rule.name(), it will be more easy to understand what the test is testing

File: server/sonar-server/src/test/java/org/sonar/server/rule/ws/UpdateActionTest.java (hunk 7 above)
Comment (gjambet) [reply server/sonar-server/src/test/java/org/sonar/server/rule/ws/UpdateActionTest.java]: done

File: server/sonar-server/src/test/java/org/sonar/server/rule/ws/UpdateActionTest.java (hunk 8 above)
Comment (gjambet) [reply server/sonar-server/src/test/java/org/sonar/server/rule/ws/UpdateActionTest.java]: done

File: server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java (hunk 4 above)
Comment (gjambet) [reply server/sonar-server/src/test/java/org/sonar/server/rule/RuleUpdaterTest.java]: i said done

**Comment by julienlancelot** [comment]
Last minute comment : name of the commit should contain the fact that the issue is when updating the description, something like "SONAR-10163 Updating description of built-in rules now fails with error 400"
